### PR TITLE
qt59.qtbase: fix darwin build so qtdeclarative can build

### PR DESCRIPTION
**Pull request by j-hao**
###### Motivation for this change

All changes are for Darwin: move sed command to patch files; disable mesa; fix OpenGL and AGL header files path. qtdeclarative and qtwebkit can be built

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Review by ttuegel — COMMENTED** [review]
I cannot test on macOS, so I should not approve this request, but neither does anything look wrong.

File: pkgs/development/libraries/qt-5/5.9/qtbase/default.nix [hunk 1]
```diff
@@ -80,7 +80,7 @@ stdenv.mkDerivation {
 
   patches =
     copyPathsToStore (lib.readPathsFromFile ./. ./series)
-    ++ stdenv.lib.optional stdenv.isDarwin ./darwin-cf.patch;
+    ++ stdenv.lib.optional stdenv.isDarwin (lib.readPathsFromFile ./. ./darwin-series);
```
Comment (LnL7) [inline pkgs/development/libraries/qt-5/5.9/qtbase/default.nix]: This doesn't seem to work correctly. I don't really get why it's done like that for the other patches either, what's the advantage over just a list of paths?

```
applying patch /nix/store/bz3c8wjkkippnb60hcqydklkpx6jngrm-git-export/pkgs/development/libraries/qt-5/5.9/qtbase/mkspecs-common-mac.patch
/nix/store/wh1ck71iabaw743xnf1j3aj9y8wh9jv3-stdenv-darwin/setup: line 740: /nix/store/bz3c8wjkkippnb60hcqydklkpx6jngrm-git-export/pkgs/development/libraries/qt-5/5.9/qtbase/mkspecs-common-mac.patch: No such file or directory
builder for ‘/nix/store/wbpddvjbjm993ffmcl5h2v7z3bfkhbm4-qtbase-5.9.1.drv’ failed with exit code 1
```

File: pkgs/development/libraries/qt-5/5.9/qtbase/default.nix [hunk 2]
```diff
@@ -45,7 +45,7 @@ stdenv.mkDerivation {
       libjpeg libpng libtiff
     ]
 
-    ++ lib.optional mesaSupported mesa
+    ++ lib.optional (mesaSupported && !stdenv.isDarwin) mesa
```
Comment (LnL7) [inline pkgs/development/libraries/qt-5/5.9/qtbase/default.nix]: This already defaults to false on darwin.

File: pkgs/development/libraries/qt-5/5.9/qtbase/default.nix (hunk 1 above)
Comment (ttuegel) [reply pkgs/development/libraries/qt-5/5.9/qtbase/default.nix]: > I don't really get why it's done like that for the other patches either, what's the advantage over just a list of paths?

I can point `quilt` at that list of patches and work with them on top of the unpacked sources (almost) as if it was Git.

File: pkgs/development/libraries/qt-5/5.9/qtbase/default.nix (hunk 2 above)
Comment (j-hao) [reply pkgs/development/libraries/qt-5/5.9/qtbase/default.nix]: On Darwin, the mesaSupported is passed in as true instead of using default value (false). That's the reason why qtdeclarative cannot build. If you look at output of qtbase configuration, the current result shows OpenVG is yes due to mesaSupported value as true.

File: pkgs/development/libraries/qt-5/5.9/qtbase/default.nix (hunk 1 above)
Comment (j-hao) [reply pkgs/development/libraries/qt-5/5.9/qtbase/default.nix]: Oh, I just follow the current way of qt* modules. They are using a list of patches even for just one patch. Like https://github.com/NixOS/nixpkgs/tree/master/pkgs/development/libraries/qt-5/5.9/qtdeclarative

@LnL7 are you sure you pulled in all the changes? Looks like the patch file is not in your test env

File: pkgs/development/libraries/qt-5/5.9/qtbase/default.nix [hunk 3]
```diff
@@ -110,10 +110,10 @@ stdenv.mkDerivation {
...
           -e 's#val=$(echo $sdk_val $(echo $val | cut -s -d \x27 \x27 -f 2-))##' \
           ./configure
-      sed -i '3,$d' ./mkspecs/features/mac/default_pre.prf
-      sed -i '27,$d' ./mkspecs/features/mac/default_post.prf
-      sed -i '1,$d' ./mkspecs/features/mac/sdk.prf
-      sed -i 's/QMAKE_LFLAGS_RPATH      = -Wl,-rpath,/QMAKE_LFLAGS_RPATH      =/' ./mkspecs/common/mac.conf
+          substituteInPlace ./mkspecs/common/mac.conf \
+              --replace "/System/Library/Frameworks/OpenGL.framework/" "${darwin.apple_sdk.frameworks.OpenGL}/Library/Frameworks/OpenGL.framework/"
+          substituteInPlace ./mkspecs/common/mac.conf \
+              --replace "/System/Library/Frameworks/AGL.framework/" "${darwin.apple_sdk.frameworks.AGL}/Library/Frameworks/AGL.framework/"
```
Comment (periklis) [inline pkgs/development/libraries/qt-5/5.9/qtbase/default.nix]: Why is this not part of the patch file?

File: pkgs/development/libraries/qt-5/5.9/qtbase/default.nix (hunk 1 above)
Comment (ttuegel) [reply pkgs/development/libraries/qt-5/5.9/qtbase/default.nix]: @j-hao You need to use `copyPathsToStore` as on line 82. This is probably broken for @LnL7 because he is using sandboxes, so his builder cannot access non-store paths.

Comment (j-hao) [reply pkgs/development/libraries/qt-5/5.9/qtbase/default.nix]: Let me fix this

File: pkgs/development/libraries/qt-5/5.9/qtbase/default.nix (hunk 3 above)
Comment (j-hao) [reply pkgs/development/libraries/qt-5/5.9/qtbase/default.nix]: Can we know the actual ${darwin.apple_sdk.frameworks.AGL} path within patch file?

File: pkgs/development/libraries/qt-5/5.9/qtbase/default.nix (hunk 1 above)
Comment (LnL7) [reply pkgs/development/libraries/qt-5/5.9/qtbase/default.nix]: Ah yes, that must be it. The hydra builds would have the same problem.

File: pkgs/development/libraries/qt-5/5.9/qtbase/default.nix (hunk 3 above)
Comment (LnL7) [reply pkgs/development/libraries/qt-5/5.9/qtbase/default.nix]: don't think so, but I think replacing paths like this preferable since it doesn't break as easily as a patch

Comment (periklis) [reply pkgs/development/libraries/qt-5/5.9/qtbase/default.nix]: Oh thanks for clarifying this. I missed the `${darwin...}` substitutions.

Comment (periklis) [reply pkgs/development/libraries/qt-5/5.9/qtbase/default.nix]: btw. these two lines will probably enable `qt5.qtwebkit`, too.

Comment (j-hao) [reply pkgs/development/libraries/qt-5/5.9/qtbase/default.nix]: Yes, I can build qt5.qtwebkit locally

**Comment by j-hao** [comment]
Updated new version. Let me know if anything need to change

**Comment by LnL7** [comment]
All of the remaining failures are caused by a chmod now, including a dependency of qtdeclarative.

```
Moving /nix/store/pkw2bp04m7c9xywjwva365glihcwrxqg-qtsvg-5.9.1/lib/qt-5.9/plugins to /nix/store/0zzl2389avjd6nwz8q7cw88xmirhzja4-qtsvg-5.9.1-bin/lib/qt-5.9/plugins
Removing empty /nix/store/pkw2bp04m7c9xywjwva365glihcwrxqg-qtsvg-5.9.1/lib/qt-5.9/ and (possibly) its parents
chmod: changing permissions of '/nix/store/pkw2bp04m7c9xywjwva365glihcwrxqg-qtsvg-5.9.1/lib/cmake/Qt5/Qt5Config.cmake': Operation not permitted
builder for ‘/nix/store/6aq90l300875x6jaimcl9avsz31p1qxz-qtsvg-5.9.1.drv’ failed with exit code 1
```

**Comment by j-hao** [comment]
@LnL7 I'm having trouble to reproduce the error you saw. Can you let me know how to reproduce it?

**Comment by LnL7** [comment]
@j-hao Hmm, maybe the minimal sandbox is only used when build user separation is enabled. Are you using the nix-daemon?

**Comment by j-hao** [comment]
@LnL7 the problem occurred if I use nix-daemon. I have switched to nix-daemon and tested the build. It can now build qt5.full for me. Please check

**Comment by LnL7** [comment]
@j-hao Nice, *everything* builds now except for qtwebengine.

**Review by LnL7 — APPROVED** [review]
